src/BinSkim.Rules/ELFRules/BA3011.EnableBindNow.cs

```diff
@@ -0,0 +1,111 @@
+            nameof(RuleResources.BA3011_Pass),
+            nameof(RuleResources.BA3011_Error),
+            nameof(RuleResources.NotApplicable_InvalidMetadata)
+        };
+
+        public override AnalysisApplicability CanAnalyzeElf(ELFBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            IELF elf = target.ELF;
+
+            if (elf.Type == FileType.Core || elf.Type == FileType.None || elf.Type == FileType.Relocatable)
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IELF elf = context.ELFBinary().ELF;
+
+            if (HasBindNowFlag(elf))
+            {
+                // Pass
+                context.Logger.Log(this,
+                RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                    nameof(RuleResources.BA3011_Pass),
+                    context.TargetUri.GetFileName()));
+            }
+            else
+            {
+                // Fail
+                context.Logger.Log(this,
+                RuleUtilities.BuildResult(FailureLevel.Error, context, null,
+                    nameof(RuleResources.BA3011_Error),
+                    context.TargetUri.GetFileName()));
+            }
+        }
+
+        // Pwntools checksec calls out numerous ways an ELF can specify full RELRO to the
+        // loader. All have the intended effect of eagerly resolving all plt functions and
+        // subsequently marking the entire GOT read-only (at least, when there
+        // is also a GNU_RELRO section)
+        // https://github.com/Gallopsled/pwntools/blob/ec9dfe108b85ac47983e9a98808fcdbb50cb0cdd/pwnlib/elf/elf.py#L1578
+        private static bool HasBindNowFlag(IELF elf)
+        {
+            try
+            {
+                var dynamicSection = (ELFSharp.ELF.Sections.IDynamicSection)elf.GetSection(".dynamic");
+                if (dynamicSection == null)
+                {
+                    return false;
+                }
+
+                foreach (ELFSharp.ELF.Sections.DynamicEntry<ulong> entry in dynamicSection.Entries)
+                {
+                    if ((entry.Tag == ELFSharp.ELF.Sections.DynamicTag.BindNow) ||
+                        (entry.Tag == ELFSharp.ELF.Sections.DynamicTag.Flags && (entry.Value & DF_BIND_NOW) != 0) ||
+                        (entry.Tag == ELFSharp.ELF.Sections.DynamicTag.Flags1 && (entry.Value & DF_1_NOW) != 0))
+                    {
+                        return true;
+                    }
+                }
+            }
+            catch (InvalidCastException)

```

Comment:
>InvalidCastException [](start = 19, length = 20)

Just curious, what's producing these exceptions?

Reply:
The project I'm working on separates executables into two files like in [this](https://stackoverflow.com/questions/46197810/separating-out-symbols-and-stripping-unneeded-symbols-at-the-same-time) post. The .dbg file contains only debug info, and therefore isn't even executable. But since it's an ELF, it gets parsed by `binskim analyze`. As a result, the output of binskim is meaningless, so I was happy to just silence the exception.

I haven't seen any other case where this happens.